### PR TITLE
Create Meeting Time Algorithm

### PR DESCRIPTION
**Pull request by austinteshuba**
Created an algorithm to find optimal meeting times to work around the attendees' schedules (where required attendees must be accommodated, and optional attendees can be ignored if needed).

Algorithm high-level overview:
1) Start with all attendees (including optional)
2) Filter the given events. Remove the events that don't share any attendees from step 1
3) Sort the events from earliest start time to latest start time.
4) Iterate through the sorted events. Merge together events to make a sorted, non-overlapping list of unavailable windows
5) Invert the unavailable windows to find available windows. Only include available windows long enough to fir the meeting request
6) a) If there are no available windows, and you didn't yet try ignoring optional attendees, do that. Return to step 1 and exclude optional attendees from consideration.
b) Otherwise, return the available windows. 

This algorithm is O(nlogn), where n is the number of events passed in.

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -15,9 +15,206 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.List;
+import java.util.HashSet;
+import java.util.ArrayList;
+import java.util.stream.Collectors;
```
Comment (gbuenoandrade) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Nit: Imported names should be in ASCII order (https://google.github.io/styleguide/javaguide.html#s3.3.3-import-ordering-and-spacing)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -15,9 +15,206 @@
...
+   * Takes in a list of busy periods, and outputs a list of available periods that are large enough
+   * to accommodate the requested event duration.
+   *
+   * @param busyTimes a sorted list (from earliest start time to latest start time) of timeRanges
+   *     that a new event cannot overlap with
+   * @param eventDuration a requested event duration. Available time ranges must be at least this
+   *     long. Value must be > 0 minutes.
+   * @return a list of TimeRanges that represent possible meeting windows throughout the day
+   */
+  private List<TimeRange> invert(List<TimeRange> busyTimes, long eventDuration) {
```
Comment (gbuenoandrade) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Maybe a bit more descriptive name for the function?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -15,9 +15,206 @@
...
+
+    // Iterate through all busy windows.
+    // End the current availability window when the next event starts, and
+    // start the next availability window when the next event ends.
+    // If the availability window is long enough to fit the meeting request,
+    // add it to the available times.
+    for (TimeRange busyTime : busyTimes) {
+      endTime = busyTime.start();
+      if (endTime - startTime >= eventDuration) {
+        availableTimes.add(TimeRange.fromStartEnd(startTime, endTime, false));
```
Comment (gbuenoandrade) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: What if you have `busyTimes = Arrays.asList(TimeRange.fromStartDuration(TIME_0800AM, DURATION_1_HOUR), TimeRange.fromStartDuration(TIME_0800AM, DURATION_30_MINUTES))`?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -15,9 +15,206 @@
...
+        // Merged Event   : |-----------------|
+        //
+        // Case 3: If the next event starts later than the previous one ends,
+        // a gap of availability is identified. Merging can now stop
+        //
+        // Prev. Event(s) : |------|
+        // Next Event     :          |-----|
+        // No merge. Move on.
+        //
+        if (compare.contains(endTime) || compare.start() == endTime) {
```
Comment (gbuenoandrade) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Nit: could you simplify this if/else block? Hint: `Math.max`

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -15,9 +15,206 @@
...
+   * earliest in the day to latest
+   *
+   * @param events list of Event objects that the meeting request must work around
+   * @param requestedAttendees list of attendees (optional and required) requested for the new
+   *     meeting
+   * @return a list of time ranges of events that must be worked around
+   */
+  private List<TimeRange> eventsToSortedTimes(
+      Collection<Event> events, Collection<String> requestedAttendees) {
+    return events.stream()
```
Comment (gbuenoandrade) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:95]: Nice use of `stream()` 😃

**Review by gbuenoandrade — COMMENTED** [review]
Good job on the algorithm 😃 

FFT: Can you simplify it? Maybe do it in a single pass after sorting? Hint: there's a very useful technique called "[line sweep](https://codeforces.com/blog/entry/20377)". (This is totally optional.)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (austinteshuba) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:95]: Thank you!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (austinteshuba) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Huh, I definitely have not been checking for this rule. 

I think I may have made the same mistakes elsewhere - I created an issue for this so I can fix this issue globally. #35

**Comment by austinteshuba** [comment]
Optimizing Algorithm to use only one sweep!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 3 above)
Comment (austinteshuba) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: A condition of using the invert function is that no events are overlapping - in this case, the events would be overlapping. 

These two events would be merged (case 2 merge) in the merge function. 

However, I'm going to be optimizing the algorithm so this function may become outdated anyways. 

Also, thanks for the edge case! I just wrote a test for this

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (austinteshuba) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Agreed, but the function is now gone since I optimized to use a single pass

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
Comment (austinteshuba) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Hmmm - I'm not so sure how to do this. 

There are three cases, each with distinct behavior (though case one and case two do both increment the counter), so I can't reduce the number of cases. Even if I moved the increment to a separate block, it's still three if blocks. Additionally, I see two checks that have to be preformed on case one no matter what: start time of the next event has to be before the merged event's end time, and the end time of the next event has to be later than the merged event's end time.

I changed the if block to be more straightforward, but I don't know how to simplify it with Math.max. Please let me know if you have any ideas!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (gbuenoandrade) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Apparently, you can try `Code | Optimize imports and Code | Rearrange Code` on IntelliJ

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -15,9 +15,206 @@
...
+        //
+        // Case 3: If the next event starts later than the previous one ends,
+        // a gap of availability is identified. Merging can now stop
+        //
+        // Prev. Event(s) : |------|
+        // Next Event     :          |-----|
+        // No merge. Move on.
+        //
+        if (compare.contains(endTime) || compare.start() == endTime) {
+          endTime = eventTimes.get(currentEventIndex + 1).end();
```
Comment (gbuenoandrade) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Nit: Use the `compare` variable you defined

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
Comment (gbuenoandrade) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: If you make `endTime = Math.max(endTime, compare.end())` you can treat case i) and ii) as being the same

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 6 above)
Comment (austinteshuba) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Oops - fixed

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
Comment (austinteshuba) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Ohhh nifty - I was only looking at the check not the content. 

Thanks!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (austinteshuba) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Cool - will do on a hotfix then

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -14,10 +14,191 @@
...
+   *
+   * @param eventTimes a sorted list of event time ranges (from earliest start time to latest start
+   *     time)
+   * @param requestedDurationMinutes a length of time given in minutes for the requested meeting.
+   *     Must be > 0.
+   * @return a sorted list of non-overlapping time ranges (from earliest start time to latest start
+   *     time)
+   */
+  private List<TimeRange> generateSuggestedTimes(List<TimeRange> eventTimes, long requestedDurationMinutes) {
+    // Create empty list to store busy periods
```
Comment (gbuenoandrade) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: 1. Nit: Did you mean available times?
2. FFT: Having comments describing what the code is doing often leads to that kind of "error" because you end up violating the DRY principle. The name of your variable, `availableTimes`, already describes what it is meant for (kudos for picking that name), and `= new ArrayList<>()` tells the reader you are creating an empty list...

**Review by gbuenoandrade — APPROVED** [review]
Great algorithm and implementation!

LGTM

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 7 above)
Comment (austinteshuba) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I agree completely - I think leaving more meaningful, sparse comments is probably a good focus for me. 

Definitely going to review the training we had on meaningful comments - often I leave these just to keep myself organized, but I should take the time to remove them so the code becomes more maintainable

**Comment by antmar** [comment]
I really like the detailed comments explaining the two cases with ASCII diagrams! It's exactly the kind of non-obvious detail that comments can help explain.
